sklearn/gaussian_process/kernels.py

```diff
@@ -0,0 +1,532 @@
+          specified via the param_space and may be optimized. The other
+          kernel parameters are set directly  at initialization and are kept
+          fixed.
+
+    Parameters
+    ----------
+    metric : string, or callable
+        The metric to use when calculating kernel between instances in a
+        feature array. If metric is a string, it must be one of the metrics
+        in pairwise.PAIRWISE_KERNEL_FUNCTIONS.
+        If metric is "precomputed", X is assumed to be a kernel matrix.
+        Alternatively, if metric is a callable function, it is called on each
+        pair of instances (rows) and the resulting value recorded. The callable
+        should take two arrays from X as input and return a value indicating
+        the distance between them.
+
+    `**kwds` : optional keyword parameters
+        Any further parameters are passed directly to the kernel function.
+    """
+
+    def __init__(self, param_space=1.0, metric="linear", **kwargs):
+        self._parse_param_space(param_space)
+        self.metric = metric
+        self.kwargs = kwargs
+        if "gamma" in kwargs:
+            raise ValueError(
+                "Gamma must not be set directly but via param_space.")
+
+    @property
+    def params(self):
+        return np.asarray([self.gamma])
+
+    @params.setter
+    def params(self, theta):
+        self.gamma = theta[0]
+
+    def __call__(self, X, Y=None, eval_gradient=False):
+        """ Return the kernel k(X, Y) and optionally its gradient.
+
+        Parameters
+        ----------
+        X : array, shape (n_samples_X, n_features)
+            Left argument of the returned kernel k(X, Y)
+
+        Y : array, shape (n_samples_Y, n_features), (optional, default=None)
+            Right argument of the returned kernel k(X, Y). If None, k(X, X)
+            if evaluated instead.
+
+        eval_gradient : bool (optional, default=False)
+            Determines whether the gradient with respect to the kernel
+            hyperparameter is determined. Only supported when Y is None.
+
+        Returns
+        -------
+        K : array, shape (n_samples_X, n_samples_Y)
+            Kernel k(X, Y)
+
+        K_gradient : array (opt.), shape (n_samples_X, n_samples_X, n_params)
+            The gradient of the kernel k(X, X) with repect to the
+            hyperparameter of the kernel. Only returned when eval_gradient
+            is True.
+        """
+        K = pairwise_kernels(X, Y, metric=self.metric, gamma=self.gamma,
+                             filter_params=True, **self.kwargs)
+        if eval_gradient:
+            # approximate gradient numerically
+            K_gradient = np.empty((K.shape[0], K.shape[1], 1))
+
+            def f(gamma, i, j):  # helper function
+                return pairwise_kernels(
+                    X, Y, metric=self.metric, gamma=gamma,
+                    filter_params=True, **self.kwargs)[i, j]
+            # XXX: avoid python for-loops
```

Comment:
Some of the optimizers in `scipy.optimize` - I think including `fmin_l_bfgs_b` - have the option to simply not pass the derivative. My intuition is that they would get the job done faster without the derivative than using a very slow one.

Wrapping a function `f(X, Y)` should probably only help the composability of the kernels. You could write a function `make_kernel(kernel_func, derivative=None, parameter_names=None)` that returns a kernel implementing that function (and derivative if available) and adds the composability. The `Kernel` class could actually take these functions in the constructor, too (but I don't see a problem in keeping this functionality isolated within `PairwiseKernels`).

Via `__rmul__` it may be able to multiply a pure function `f(X, Y)` with a kernel and obtain a kernel, but that is probably not so useful, since there would never be a derivative for the ones constructed that way. (It depends on how fast optimization is without the derivative.)

The idea behind my comments is to make kernels interchangeably usable by all methods that reportedly use them, e.g. `KernelRidge`. The fact that now the kernel operation is accessible with `__call__` should take care of this, but a test may be in order.